File: src/test/java/pl/training/module7/BreakMethodEquivalenceTest.java
package pl.training.module7;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

import org.junit.jupiter.api.Test;

import pl.training.module7.breakmethod.ManifestEntry;
import pl.training.module7.breakmethod.after.ReleaseManifestBuilder;
import pl.training.module7.breakmethod.before.LegacyReleaseManifestBuilder;

final class BreakMethodEquivalenceTest {
    @Test
    void preservesOrderingRenderingAndTheCallersCollection() {
        var entries = new ArrayList<>(List.of(
                new ManifestEntry("web", "sha-web", 20),
                new ManifestEntry("worker", "sha-worker", 10),
                new ManifestEntry("api", "sha-api", 10)));
        List<ManifestEntry> originalOrder = List.copyOf(entries);

        var legacy = new LegacyReleaseManifestBuilder();
        var refactored = new ReleaseManifestBuilder();

        String expected = "10|api|sha-api\n"
                + "10|worker|sha-worker\n"
                + "20|web|sha-web";

        assertEquals(expected, legacy.build(entries));
        assertEquals(expected, refactored.build(entries));
        assertEquals(originalOrder, entries);
        assertEquals("", legacy.build(List.of()));
        assertEquals("", refactored.build(List.of()));
    }

    @Test
    void preservesValidationTypeMessageAndEncounterOrder() {
        var legacy = new LegacyReleaseManifestBuilder();
        var refactored = new ReleaseManifestBuilder();

        assertSameFailure(
                () -> legacy.build(null),
                () -> refactored.build(null));
        assertSameFailure(
                () -> legacy.build(listContainingNull()),
                () -> refactored.build(listContainingNull()));
        assertSameFailure(
                () -> legacy.build(List.of(new ManifestEntry(null, null, -1))),
                () -> refactored.build(List.of(new ManifestEntry(null, null, -1))));
        assertSameFailure(
                () -> legacy.build(List.of(new ManifestEntry(" ", null, -1))),
                () -> refactored.build(List.of(new ManifestEntry(" ", null, -1))));
        assertSameFailure(
                () -> legacy.build(List.of(new ManifestEntry("api", null, -1))),
                () -> refactored.build(List.of(new ManifestEntry("api", null, -1))));
        assertSameFailure(
                () -> legacy.build(List.of(new ManifestEntry("api", " ", -1))),
                () -> refactored.build(List.of(new ManifestEntry("api", " ", -1))));
        assertSameFailure(
                () -> legacy.build(List.of(new ManifestEntry("api", "sha", -1))),
                () -> refactored.build(List.of(new ManifestEntry("api", "sha", -1))));
    }

    @Test
    void preservesInputOrderWhenAllSortKeysAreEqual() {
        List<ManifestEntry> entries = List.of(
                new ManifestEntry("api", "sha-first", 10),
                new ManifestEntry("api", "sha-second", 10));
        String expected = "10|api|sha-first\n10|api|sha-second";

        assertEquals(expected, new LegacyReleaseManifestBuilder().build(entries));
        assertEquals(expected, new ReleaseManifestBuilder().build(entries));
    }

    private static List<ManifestEntry> listContainingNull() {
        var entries = new ArrayList<ManifestEntry>();
        entries.add(null);
        return entries;
    }

    private static void assertSameFailure(
            Supplier<String> legacyCall,
            Supplier<String> refactoredCall) {

        RuntimeException legacyFailure = assertThrows(
                RuntimeException.class,
                legacyCall::get);
        RuntimeException refactoredFailure = assertThrows(
                RuntimeException.class,
                refactoredCall::get);

        assertEquals(legacyFailure.getClass(), refactoredFailure.getClass());
        assertEquals(legacyFailure.getMessage(), refactoredFailure.getMessage());
    }
}

File: csharp/tests/Training.Module7.Tests/BreakMethodEquivalenceTest.cs
using Training.Module7.BreakMethod;
using Training.Module7.BreakMethod.After;
using Training.Module7.BreakMethod.Before;
using static Training.Module7.Tests.FailureAssertions;

namespace Training.Module7.Tests;

public sealed class BreakMethodEquivalenceTest
{
    [Fact]
    public void PreservesOrderingRenderingAndTheCallersCollection()
    {
        var entries = new List<ManifestEntry>
        {
            new("web", "sha-web", 20),
            new("worker", "sha-worker", 10),
            new("api", "sha-api", 10)
        };
        var originalOrder = entries.ToList();

        var legacy = new LegacyReleaseManifestBuilder();
        var refactored = new ReleaseManifestBuilder();

        const string expected = "10|api|sha-api\n"
            + "10|worker|sha-worker\n"
            + "20|web|sha-web";

        Assert.Equal(expected, legacy.Build(entries));
        Assert.Equal(expected, refactored.Build(entries));
        Assert.Equal(originalOrder, entries);
        Assert.Equal("", legacy.Build([]));
        Assert.Equal("", refactored.Build([]));
    }

    [Fact]
    public void PreservesValidationTypeMessageAndEncounterOrder()
    {
        var legacy = new LegacyReleaseManifestBuilder();
        var refactored = new ReleaseManifestBuilder();

        AssertSameFailure(
            () => legacy.Build(null!),
            () => refactored.Build(null!));
        AssertSameFailure(
            () => legacy.Build(ListContainingNull()),
            () => refactored.Build(ListContainingNull()));
        AssertSameFailure(
            () => legacy.Build([new ManifestEntry(null!, null!, -1)]),
            () => refactored.Build([new ManifestEntry(null!, null!, -1)]));
        AssertSameFailure(
            () => legacy.Build([new ManifestEntry(" ", null!, -1)]),
            () => refactored.Build([new ManifestEntry(" ", null!, -1)]));
        AssertSameFailure(
            () => legacy.Build([new ManifestEntry("api", null!, -1)]),
            () => refactored.Build([new ManifestEntry("api", null!, -1)]));
        AssertSameFailure(
            () => legacy.Build([new ManifestEntry("api", " ", -1)]),
            () => refactored.Build([new ManifestEntry("api", " ", -1)]));
        AssertSameFailure(
            () => legacy.Build([new ManifestEntry("api", "sha", -1)]),
            () => refactored.Build([new ManifestEntry("api", "sha", -1)]));
    }

    [Fact]
    public void PreservesInputOrderWhenAllSortKeysAreEqual()
    {
        ManifestEntry[] entries =
        [
            new("api", "sha-first", 10),
            new("api", "sha-second", 10)
        ];
        const string expected = "10|api|sha-first\n10|api|sha-second";

        Assert.Equal(expected, new LegacyReleaseManifestBuilder().Build(entries));
        Assert.Equal(expected, new ReleaseManifestBuilder().Build(entries));
    }

    private static List<ManifestEntry> ListContainingNull() => [null!];
}
